### ISSUE TITLE
Assertion Failed due to invalid bytes in instruction stream

### ISSUE THREAD

**Issue by greenozon** [icsharpcode/ILSpy#1179]
ILSpy version ILSpy version 3.2.0.3817-beta

issue
```
Error decompiling System.Void BabelObfuscatorAttribute::.ctor()
 ---> ICSharpCode.ILSpy.AssertionFailedException: 
   at ICSharpCode.ILSpy.ILSpyTraceListener.Fail(String message, String detailMessage) in offset 380
   at ICSharpCode.ILSpy.ILSpyTraceListener.Fail(String message) in offset 9
   at System.Diagnostics.TraceInternal.Fail(String message) in offset 55
   at System.Diagnostics.Debug.Assert(Boolean condition) in offset 6
   at ICSharpCode.Decompiler.IL.ILReader.ReadInstructions(CancellationToken cancellationToken) in offset 701
   at ICSharpCode.Decompiler.IL.ILReader.ReadIL(MethodBody body, CancellationToken cancellationToken) in offset 43
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.DecompileBody(MethodDefinition methodDefinition, IMethod method, EntityDeclaration entityDecl, DecompileRun decompileRun, ITypeResolveContext decompilationContext) in offset 54
-- continuing with outer exception (ICSharpCode.Decompiler.DecompilerException) --
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.DecompileBody(MethodDefinition methodDefinition, IMethod method, EntityDeclaration entityDecl, DecompileRun decompileRun, ITypeResolveContext decompilationContext) in offset 914
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.DoDecompile(MethodDefinition methodDefinition, IMethod method, DecompileRun decompileRun, ITypeResolveContext decompilationContext) in offset 126
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.Decompile(IList`1 definitions) in offset 383
   at ICSharpCode.ILSpy.CSharpLanguage.DecompileMethod(MethodDefinition method, ITextOutput output, DecompilationOptions options) in offset 141
   at ICSharpCode.ILSpy.TreeNodes.MethodTreeNode.Decompile(Language language, ITextOutput output, DecompilationOptions options) in offset 15
   at ICSharpCode.ILSpy.TextView.DecompilerTextView.DecompileNodes(DecompilationContext context, ITextOutput textOutput) in offset 69
   at ICSharpCode.ILSpy.TextView.DecompilerTextView.<>c__DisplayClass33_0.<DecompileAsync>b__0() in offset 38
```

IL code as shown by ILSpy:

```
.method public hidebysig specialname rtspecialname 
	instance void .ctor () cil managed 
{
	// Method begins at RVA 0x2050
	// Code size 14 (0xe)
	.maxstack 8

	IL_0000: br

	IL_0005: nop

	IL_000a: nop
	IL_000b: nop
	IL_000c: stloc.0
	IL_000d: ret
}
```

but
IL code as shown by dnSpy:

```
// Token: 0x06000001 RID: 1 RVA: 0x00002050 File Offset: 0x00000250
.method public hidebysig specialname rtspecialname 
	instance void .ctor () cil managed 
{
	// Header Size: 1 byte
	// Code Size: 14 (0xE) bytes
	.maxstack 8

	/* 0x00000251 3802000000   */ IL_0000: br        IL_0007

	/* 0x00000256 AB           */ IL_0005: UNKNOWN1
	/* 0x00000257 BF           */ IL_0006: UNKNOWN1

	/* 0x00000258 02           */ IL_0007: ldarg.0
	/* 0x00000259 281C00000A   */ IL_0008: call      instance void [mscorlib]System.Attribute::.ctor()
	/* 0x0000025E 2A           */ IL_000D: ret
}
```

**Comment by siegfriedpammer** [comment]
This will be fixed in ILSpy 4.0.

**Comment by greenozon** [comment]
Thanks
do you mean that v4 will you newer version of Mono.Cecil?
more robust and so on..

**Comment by greenozon** [comment]
I'm interesting cause right now on some obfuscated methods ILSpy even can't show IL !
instead it says

```
ystem.IndexOutOfRangeException: Index was outside the bounds of the array.
   at Mono.Cecil.Cil.CodeReader.ReadOpCode()
   at Mono.Cecil.Cil.CodeReader.ReadCode()
   at Mono.Cecil.Cil.CodeReader.ReadFatMethod()
   at Mono.Cecil.Cil.CodeReader.ReadMethodBody()
   at Mono.Cecil.Cil.CodeReader.ReadMethodBody(MethodDefinition method)
   at Mono.Cecil.MethodDefinition.<>c.<get_Body>b__41_0(MethodDefinition method, MetadataReader reader)
   at Mono.Cecil.ModuleDefinition.Read[TItem,TRet](TRet& variable, TItem item, Func`3 read)
   at Mono.Cecil.MethodDefinition.get_Body()
   at ICSharpCode.Decompiler.Disassembler.ReflectionDisassembler.DisassembleMethodBlock(MethodDefinition method)
   at ICSharpCode.Decompiler.Disassembler.ReflectionDisassembler.DisassembleMethod(MethodDefinition method)
.....
```

**Comment by siegfriedpammer** [comment]
We are in the process of moving to System.Reflection.Metadata... not sure if that will fix problems with obfuscated assemblies, but it will enable us to not crash on invalid IL.

**Comment by siegfriedpammer** [comment]
Could you please provide the assembly for testing purposes? Thank you very much!

**Comment by greenozon** [comment]
yeah, np - https://www.sendspace.com/file/ht469k

during searching for this guy I've found another issues :) 
will open up new tickets in a little while

**Comment by dgrunwald** [comment]
The exceptions on invalid opcode bytes are fixed in ILSpy 4.0 (which just merged to master).
